pkg/distributor/distributor.go

```diff
@@ -481,16 +495,79 @@ func (d *Distributor) sendRequests(ctx context.Context, req *distributormodel.Pu
 			d.sendProfiles(localCtx, ingester, samples, &tracker)
 		}(ingesterDescs[ingester], samples)
 	}
-	select {
-	case err = <-tracker.err:
-		return nil, err
-	case <-tracker.done:
-		return connect.NewResponse(&pushv1.PushResponse{}), nil
-	case <-ctx.Done():
-		return nil, ctx.Err()
+	numRetries := 0
+	maxRetries := 1
+	for {
+		select {
+		case err = <-tracker.err:
+			// retry failed profiles using fallback nodes once
+			if numRetries < maxRetries {
+				return nil, err
+			}
+			samplesByIngester = map[string][]*profileTracker{}
+			ingesterDescs = map[string]ring.InstanceDesc{}
+			pending := int32(0)
+			for i, p := range profiles {
+				if p.failed.Load() > 0 {
+					if len(p.fallbackNodes) == 0 {
+						level.Warn(d.logger).Log("msg", "no fallback nodes to send failed profile to", "err", err)
+						continue
+					}
+					ingester := p.fallbackNodes[0]
+					samplesByIngester[ingester.Addr] = append(samplesByIngester[ingester.Addr], profiles[i])
+					ingesterDescs[ingester.Addr] = ingester
+					pending++
+				}
+			}
+			if pending > 0 {
+				tracker.samplesPending.Store(pending)
+				numRetries++
+				for ingester, samples := range samplesByIngester {
+					go func(ingester ring.InstanceDesc, samples []*profileTracker) {
+						// Use a background context to make sure all ingesters get samples even if we return early
+						localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.PushTimeout)
+						defer cancel()
+						localCtx = tenant.InjectTenantID(localCtx, tenantID)
+						if sp := opentracing.SpanFromContext(ctx); sp != nil {
+							localCtx = opentracing.ContextWithSpan(localCtx, sp)
+						}
+						d.sendProfiles(localCtx, ingester, samples, &tracker)
+					}(ingesterDescs[ingester], samples)
+				}
+			} else {
+				level.Warn(d.logger).Log("msg", "error received but no profiles marked as failed", "err", err)
+			}
+		case <-tracker.done:
+			return connect.NewResponse(&pushv1.PushResponse{}), nil
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
 	}
 }
 
+func getShard(instance *ring.InstanceDesc, key uint32, logger log.Logger) uint32 {
+	parts := strings.Split(instance.Id, "-")
+	if len(parts) < 2 {
+		level.Warn(logger).Log("msg", "cannot determine shard, unexpected instance id", "instance_id", instance.Id)
+		return 0
+	}
+	instanceIdNum, err := strconv.Atoi(parts[len(parts)-1])
+	if err != nil {
+		level.Warn(logger).Log("msg", "cannot determine shard, unexpected instance id", "instance_id", instance.Id)
+		return 0
+	}
+	i, found := slices.BinarySearch(instance.Tokens, key)
+	if found {
+		i = i + 1
+	}
+	if i >= len(instance.Tokens) {
+		i = 0
+	}
+	shard := uint32(instanceIdNum*len(instance.Tokens) + i)
+	level.Debug(logger).Log("msg", "shard found", "shard", shard)
+	return shard
+}
+
```

Comment:
I'm wondering if we could get it via:
```
shard_id = hash(tenant_id+service_name) % (instances_total * shards_per_instance)
```

Shard placement does not have to match the ring distribution (maybe event should not)

Reply:
The current logic will result in more even block sizes when we have e.e., a single large tenant service because it'll be spread across multiple shards as opposed to a single one. We could solve that by spreading it to more ingesters though in the future, so I can simplify things.

Reply:
Sorry, didn't mean to confuse you. I'm just thinking out loud: the current implementation is fine – we can rely on the instance names in the prototype. In real life, we want to have something more abstract, though.

You're right about shard distribution. I'm just wondering if this is bad: this allows us to scope the app to a single shard (which drastically simplifies deduplication), while distributing the load among ingesters. The block size is determined by the compaction strategy, so we have control over this. The method is bad for another reason: really bad remapping when adding or removing instances. There are workarounds, but I don't think they are any better than rendezvous hashing (HRW) or jump hashing – something we will experiment with later on.

BTW, you can also determine the shard based on the _first_ instance: `hash(instance) % shards_total` :) Although this is also prone to issues with remapping.

Reply:
Will leave it as is (`hash(tenant_id+service_name) % shards_total`), we can provide alternative sharding strategies later on.